File: memcontainers/shcore/src/testexpr.rs
//! The `test` / `[` conditional-expression evaluator.
//!
//! Shared by the `test`/`[` builtin (`exec.rs`) and the `/bin/test` tool
//! (`programs/`), so the operators, precedence, and arity rules live once (B2 in
//! spirit). File metadata is injected as a `stat` closure, so the grammar itself
//! stays OS-free and unit-testable; the builtin passes a closure over `ShellOs::stat`
//! and the `/bin` tool one over the raw syscall.
//!
//! Grammar: `or → and → factor`, with `!` negation, `( … )` grouping, and `-a`
//! binding tighter than `-o`. POSIX one-argument form is a plain non-null string test
//! even when the token looks like an operator.

use alloc::format;
use alloc::string::{String, ToString};
use alloc::vec::Vec;

use crate::os::FileStat;

/// File metadata source for the file-test primaries (`-e`/`-f`/`-d`/…). `None` ⇒ the
/// path does not exist or cannot be stat'd.
pub type Stat<'a> = dyn FnMut(&str) -> Option<FileStat> + 'a;

/// Evaluate a `test`/`[` invocation. `name` is `"test"` or `"["`; `args` is the
/// operands (for `[`, the trailing `]` is required and stripped here). Returns the
/// truth value, or an already-prefixed error message ready to print to stderr (exit 2).
pub fn eval(name: &str, args: &[String], stat: &mut Stat) -> Result<bool, String> {
    let mut a: Vec<&str> = args.iter().map(|s| s.as_str()).collect();
    if name == "[" {
        if a.last() == Some(&"]") {
            a.pop();
        } else {
            return Err("[: missing ']'".to_string());
        }
    }
    eval_expr(&a, stat).map_err(|m| format!("test: {m}"))
}

fn eval_expr(a: &[&str], stat: &mut Stat) -> Result<bool, String> {
    if a.is_empty() {
        return Ok(false);
    }
    // POSIX's one-argument form is a plain non-null string test, even when the token
    // looks like an operator (`test !`, `test -z`, `test '('`).
    if a.len() == 1 {
        return Ok(!a[0].is_empty());
    }
    let mut i = 0usize;
    let v = test_or(a, &mut i, stat)?;
    if i != a.len() {
        return Err("too many arguments".to_string());
    }
    Ok(v)
}

fn test_or(a: &[&str], i: &mut usize, stat: &mut Stat) -> Result<bool, String> {
    let mut v = test_and(a, i, stat)?;
    while a.get(*i) == Some(&"-o") {
        *i += 1;
        let r = test_and(a, i, stat)?;
        v = v || r;
    }
    Ok(v)
}

fn test_and(a: &[&str], i: &mut usize, stat: &mut Stat) -> Result<bool, String> {
    let mut v = test_factor(a, i, stat)?;
    while a.get(*i) == Some(&"-a") {
        *i += 1;
        let r = test_factor(a, i, stat)?;
        v = v && r;
    }
    Ok(v)
}

fn test_factor(a: &[&str], i: &mut usize, stat: &mut Stat) -> Result<bool, String> {
    let tok = *a.get(*i).ok_or_else(|| "missing argument".to_string())?;
    if tok == "!" {
        *i += 1;
        return Ok(!test_factor(a, i, stat)?);
    }
    if tok == "(" {
        *i += 1;
        let v = test_or(a, i, stat)?;
        if a.get(*i) != Some(&")") {
            return Err("missing ')'".to_string());
        }
        *i += 1;
        return Ok(v);
    }
    let rem = a.len() - *i;
    if rem >= 3 && is_binop(a[*i + 1]) {
        let r = test_binary(a[*i], a[*i + 1], a[*i + 2], stat)?;
        *i += 3;
        return Ok(r);
    }
    if rem >= 2 && is_unop(a[*i]) {
        let r = test_unary(a[*i], a[*i + 1], stat)?;
        *i += 2;
        return Ok(r);
    }
    *i += 1;
    Ok(!tok.is_empty())
}

fn test_unary(op: &str, x: &str, stat: &mut Stat) -> Result<bool, String> {
    Ok(match op {
        "-z" => x.is_empty(),
        "-n" => !x.is_empty(),
        "-e" => stat(x).is_some(),
        "-f" => stat(x).map(|s| !s.is_dir).unwrap_or(false),
        "-d" => stat(x).map(|s| s.is_dir).unwrap_or(false),
        "-s" => stat(x).map(|s| s.size > 0).unwrap_or(false),
        "-r" => stat(x).map(|s| s.readable()).unwrap_or(false),
        "-w" => stat(x).map(|s| s.writable()).unwrap_or(false),
        "-x" => stat(x).map(|s| s.executable()).unwrap_or(false),
        _ => return Err(format!("unknown unary operator {op}")),
    })
}

fn test_binary(x: &str, op: &str, y: &str, stat: &mut Stat) -> Result<bool, String> {
    Ok(match op {
        "=" | "==" => x == y,
        "!=" => x != y,
        "-eq" => num(x)? == num(y)?,
        "-ne" => num(x)? != num(y)?,
        "-lt" => num(x)? < num(y)?,
        "-le" => num(x)? <= num(y)?,
        "-gt" => num(x)? > num(y)?,
        "-ge" => num(x)? >= num(y)?,
        "-nt" => {
            let mx = stat(x).map(|s| s.mtime);
            let my = stat(y).map(|s| s.mtime);
            match (mx, my) {
                (Some(a), Some(b)) => a > b,
                (Some(_), None) => true,
                _ => false,
            }
        }
        "-ot" => {
            let mx = stat(x).map(|s| s.mtime);
            let my = stat(y).map(|s| s.mtime);
            match (mx, my) {
                (Some(a), Some(b)) => a < b,
                (None, Some(_)) => true,
                _ => false,
            }
        }
        _ => return Err(format!("unknown binary operator {op}")),
    })
}

fn num(s: &str) -> Result<i64, String> {
    s.trim()
        .parse::<i64>()
        .map_err(|_| format!("integer expected: {s}"))
}

/// `test` binary operators (keep identical to the `/bin/test` tool).
pub fn is_binop(op: &str) -> bool {
    matches!(
        op,
        "=" | "==" | "!=" | "-eq" | "-ne" | "-lt" | "-le" | "-gt" | "-ge" | "-nt" | "-ot"
    )
}

/// `test` unary operators (keep identical to the `/bin/test` tool).
pub fn is_unop(op: &str) -> bool {
    matches!(
        op,
        "-z" | "-n" | "-e" | "-f" | "-d" | "-s" | "-r" | "-w" | "-x"
    )
}

#[cfg(test)]
mod tests {
    use super::*;
    use alloc::vec;

    // A fake filesystem: "f" is a 10-byte file, "d" is a directory, anything else
    // is absent. Permission bits are rwx for the owner.
    fn fake(path: &str) -> Option<FileStat> {
        match path {
            "f" => Some(FileStat { is_dir: false, size: 10, mode: 0o700, mtime: 100 }),
            "d" => Some(FileStat { is_dir: true, size: 0, mode: 0o700, mtime: 200 }),
            "empty" => Some(FileStat { is_dir: false, size: 0, mode: 0o000, mtime: 50 }),
            _ => None,
        }
    }

    fn ev(name: &str, args: &[&str]) -> Result<bool, String> {
        let a: Vec<String> = args.iter().map(|s| s.to_string()).collect();
        let mut st = fake;
        eval(name, &a, &mut st)
    }

    #[test]
    fn string_forms() {
        assert_eq!(ev("test", &["abc"]), Ok(true));
        assert_eq!(ev("test", &[""]), Ok(false));
        assert_eq!(ev("test", &["-z", ""]), Ok(true));
        assert_eq!(ev("test", &["-n", "x"]), Ok(true));
        assert_eq!(ev("test", &["a", "=", "a"]), Ok(true));
        assert_eq!(ev("test", &["a", "!=", "b"]), Ok(true));
    }

    #[test]
    fn integer_comparisons() {
        assert_eq!(ev("test", &["3", "-eq", "3"]), Ok(true));
        assert_eq!(ev("test", &["3", "-lt", "5"]), Ok(true));
        assert_eq!(ev("test", &["5", "-ge", "5"]), Ok(true));
        assert!(ev("test", &["x", "-eq", "3"]).is_err());
    }

    #[test]
    fn file_tests_via_stat() {
        assert_eq!(ev("test", &["-e", "f"]), Ok(true));
        assert_eq!(ev("test", &["-e", "nope"]), Ok(false));
        assert_eq!(ev("test", &["-f", "f"]), Ok(true));
        assert_eq!(ev("test", &["-d", "d"]), Ok(true));
        assert_eq!(ev("test", &["-d", "f"]), Ok(false));
        assert_eq!(ev("test", &["-s", "f"]), Ok(true));
        assert_eq!(ev("test", &["-s", "empty"]), Ok(false));
        assert_eq!(ev("test", &["f", "-nt", "d"]), Ok(false)); // f mtime 100 < d 200
        assert_eq!(ev("test", &["d", "-nt", "f"]), Ok(true));
    }

    #[test]
    fn connectives_negation_grouping() {
        assert_eq!(ev("test", &["a", "=", "a", "-a", "b", "=", "b"]), Ok(true));
        assert_eq!(ev("test", &["a", "=", "b", "-o", "c", "=", "c"]), Ok(true));
        assert_eq!(ev("test", &["!", "-z", "x"]), Ok(true));
        assert_eq!(ev("test", &["(", "a", "=", "a", ")"]), Ok(true));
    }

    #[test]
    fn bracket_requires_close_and_errors_prefix() {
        assert_eq!(ev("[", &["x", "]"]), Ok(true));
        assert_eq!(ev("[", &["x"]), Err("[: missing ']'".to_string()));
        assert_eq!(
            ev("test", &["a", "b", "c", "d"]),
            Err("test: too many arguments".to_string())
        );
        let _ = vec![0u8];
    }
}

File: memcontainers/shcore/src/testexpr.zig
//! test/[ expression evaluator.

const std = @import("std");
const os = @import("os.zig");

pub const EvalError = union(enum) {
    missing_bracket,
    missing_argument,
    missing_rparen,
    too_many_arguments,
    unknown_unary: []const u8,
    unknown_binary: []const u8,
    integer_expected: []const u8,
};

pub const Result = union(enum) {
    ok: bool,
    err: EvalError,
};

pub const StatFn = *const fn (*anyopaque, []const u8) ?os.FileStat;

pub fn eval(name: []const u8, args: []const []const u8, stat_ptr: *anyopaque, stat_fn: StatFn) Result {
    var end = args.len;
    if (std.mem.eql(u8, name, "[")) {
        if (end == 0 or !std.mem.eql(u8, args[end - 1], "]")) {
            return .{ .err = .missing_bracket };
        }
        end -= 1;
    }
    return evalExpr(args[0..end], stat_ptr, stat_fn);
}

fn evalExpr(args: []const []const u8, stat_ptr: *anyopaque, stat_fn: StatFn) Result {
    if (args.len == 0) return .{ .ok = false };
    if (args.len == 1) return .{ .ok = args[0].len != 0 };

    var i: usize = 0;
    const value = testOr(args, &i, stat_ptr, stat_fn);
    switch (value) {
        .err => return value,
        .ok => |v| {
            if (i != args.len) return .{ .err = .too_many_arguments };
            return .{ .ok = v };
        },
    }
}

fn testOr(args: []const []const u8, i: *usize, stat_ptr: *anyopaque, stat_fn: StatFn) Result {
    var value = testAnd(args, i, stat_ptr, stat_fn);
    while (i.* < args.len and std.mem.eql(u8, args[i.*], "-o")) {
        i.* += 1;
        const rhs = testAnd(args, i, stat_ptr, stat_fn);
        value = boolOr(value, rhs);
    }
    return value;
}

fn testAnd(args: []const []const u8, i: *usize, stat_ptr: *anyopaque, stat_fn: StatFn) Result {
    var value = testFactor(args, i, stat_ptr, stat_fn);
    while (i.* < args.len and std.mem.eql(u8, args[i.*], "-a")) {
        i.* += 1;
        const rhs = testFactor(args, i, stat_ptr, stat_fn);
        value = boolAnd(value, rhs);
    }
    return value;
}

fn testFactor(args: []const []const u8, i: *usize, stat_ptr: *anyopaque, stat_fn: StatFn) Result {
    if (i.* >= args.len) return .{ .err = .missing_argument };
    const tok = args[i.*];
    if (std.mem.eql(u8, tok, "!")) {
        i.* += 1;
        return boolNot(testFactor(args, i, stat_ptr, stat_fn));
    }
    if (std.mem.eql(u8, tok, "(")) {
        i.* += 1;
        const value = testOr(args, i, stat_ptr, stat_fn);
        if (i.* >= args.len or !std.mem.eql(u8, args[i.*], ")")) {
            return .{ .err = .missing_rparen };
        }
        i.* += 1;
        return value;
    }

    const rem = args.len - i.*;
    if (rem >= 3 and isBinop(args[i.* + 1])) {
        const value = testBinary(args[i.*], args[i.* + 1], args[i.* + 2], stat_ptr, stat_fn);
        i.* += 3;
        return value;
    }
    if (rem >= 2 and isUnop(args[i.*])) {
        const value = testUnary(args[i.*], args[i.* + 1], stat_ptr, stat_fn);
        i.* += 2;
        return value;
    }
    i.* += 1;
    return .{ .ok = tok.len != 0 };
}

fn testUnary(op: []const u8, x: []const u8, stat_ptr: *anyopaque, stat_fn: StatFn) Result {
    if (std.mem.eql(u8, op, "-z")) return .{ .ok = x.len == 0 };
    if (std.mem.eql(u8, op, "-n")) return .{ .ok = x.len != 0 };
    const st = stat_fn(stat_ptr, x);
    if (std.mem.eql(u8, op, "-e")) return .{ .ok = st != null };
    if (std.mem.eql(u8, op, "-f")) return .{ .ok = if (st) |s| !s.is_dir else false };
    if (std.mem.eql(u8, op, "-d")) return .{ .ok = if (st) |s| s.is_dir else false };
    if (std.mem.eql(u8, op, "-s")) return .{ .ok = if (st) |s| s.size > 0 else false };
    if (std.mem.eql(u8, op, "-r")) return .{ .ok = if (st) |s| s.readable() else false };
    if (std.mem.eql(u8, op, "-w")) return .{ .ok = if (st) |s| s.writable() else false };
    if (std.mem.eql(u8, op, "-x")) return .{ .ok = if (st) |s| s.executable() else false };
    return .{ .err = .{ .unknown_unary = op } };
}

fn testBinary(x: []const u8, op: []const u8, y: []const u8, stat_ptr: *anyopaque, stat_fn: StatFn) Result {
    if (std.mem.eql(u8, op, "=") or std.mem.eql(u8, op, "==")) return .{ .ok = std.mem.eql(u8, x, y) };
    if (std.mem.eql(u8, op, "!=")) return .{ .ok = !std.mem.eql(u8, x, y) };
    if (std.mem.eql(u8, op, "-eq")) return intCompare(x, y, .eq);
    if (std.mem.eql(u8, op, "-ne")) return intCompare(x, y, .ne);
    if (std.mem.eql(u8, op, "-lt")) return intCompare(x, y, .lt);
    if (std.mem.eql(u8, op, "-le")) return intCompare(x, y, .le);
    if (std.mem.eql(u8, op, "-gt")) return intCompare(x, y, .gt);
    if (std.mem.eql(u8, op, "-ge")) return intCompare(x, y, .ge);
    if (std.mem.eql(u8, op, "-nt")) {
        const mx = if (stat_fn(stat_ptr, x)) |s| s.mtime else null;
        const my = if (stat_fn(stat_ptr, y)) |s| s.mtime else null;
        return .{ .ok = if (mx) |a| if (my) |b| a > b else true else false };
    }
    if (std.mem.eql(u8, op, "-ot")) {
        const mx = if (stat_fn(stat_ptr, x)) |s| s.mtime else null;
        const my = if (stat_fn(stat_ptr, y)) |s| s.mtime else null;
        return .{ .ok = if (mx) |a| if (my) |b| a < b else false else my != null };
    }
    return .{ .err = .{ .unknown_binary = op } };
}

const Cmp = enum { eq, ne, lt, le, gt, ge };

fn intCompare(x: []const u8, y: []const u8, cmp: Cmp) Result {
    const xv = parseInt(x) orelse return .{ .err = .{ .integer_expected = x } };
    const yv = parseInt(y) orelse return .{ .err = .{ .integer_expected = y } };
    return .{ .ok = switch (cmp) {
        .eq => xv == yv,
        .ne => xv != yv,
        .lt => xv < yv,
        .le => xv <= yv,
        .gt => xv > yv,
        .ge => xv >= yv,
    } };
}

fn parseInt(s: []const u8) ?i64 {
    return std.fmt.parseInt(i64, std.mem.trim(u8, s, " \t\r\n"), 10) catch null;
}

pub fn isBinop(op: []const u8) bool {
    inline for (.{ "=", "==", "!=", "-eq", "-ne", "-lt", "-le", "-gt", "-ge", "-nt", "-ot" }) |candidate| {
        if (std.mem.eql(u8, op, candidate)) return true;
    }
    return false;
}

pub fn isUnop(op: []const u8) bool {
    inline for (.{ "-z", "-n", "-e", "-f", "-d", "-s", "-r", "-w", "-x" }) |candidate| {
        if (std.mem.eql(u8, op, candidate)) return true;
    }
    return false;
}

fn boolNot(r: Result) Result {
    return switch (r) {
        .ok => |v| .{ .ok = !v },
        .err => r,
    };
}

fn boolAnd(a: Result, b: Result) Result {
    return switch (a) {
        .err => a,
        .ok => |av| switch (b) {
            .err => b,
            .ok => |bv| .{ .ok = av and bv },
        },
    };
}

fn boolOr(a: Result, b: Result) Result {
    return switch (a) {
        .err => a,
        .ok => |av| switch (b) {
            .err => b,
            .ok => |bv| .{ .ok = av or bv },
        },
    };
}
